04-data-structures-part1.Rmd

```diff
@@ -2,336 +2,271 @@
-> ```
->
-
-This is called implicit coercion.
-
-The coercion rule goes `logical` -> `integer` -> `numeric` -> `complex` ->
-`character`.
-
-You can also coerce vectors explicitly using the `as.<class_name>`. Example
+Given what we've learned so far, what do you think the following will produce?
 
 ```{r}
-as.numeric()
-as.character()
+x <- c(2,6,'3')
 ```
 
-R will try to do whatever makes the most sense for that value:
+This is something called *type coercion*, and it is the source of many surprises and the reason why we need to be aware of the basic data types and how R will interpret them. Consider:
 
 ```{r}
-as.character(x)
+x <- c('a', TRUE)
+x
+x <- c(0, TRUE)
+x
 ```
 
-```{r}
-as.complex(x)
-```
+The coercion rules go: `logical` -> `integer` -> `numeric` -> `complex` -> `character`. You can try to force coercion against this flow using the `as.` functions:
 
 ```{r}
-x <- 0:6
-as.logical(x)
+x <- c('0','2','4')
+x
+y <- as.numeric(x)
+y
+z <- as.logical(y)
+z
 ```
 
-This is behaviour you will find in many programming languages. 0 is
-FALSE, while every other number is treated as TRUE.
-Sometimes coercions, especially nonsensical ones won't work.
+As you can see, some surprising things can happen when R forces one basic data type into another! Nitty-gritty of type coercion aside, the point is: if your data doesn't look like what you thought it was going to look like, type coercion may well be to blame; make sure everything is the same type in your vectors and your columns of data.frames, or you will get nasty surprises!
 
-In some cases, R won't be able to do anything sensible:
+Concatenate will also append things to an existing vector:
 
 ```{r}
-x <- c("a", "b", "c")
-as.numeric(x)
-as.logical(x)
+x <- c('a', 'b')
+x
+x <- c(x, 'SWC')
+x
 ```
 
-In both cases, a vector of "NAs" was returned, and in the first case
-so was a warning.
-
-> ## Tip: Special Objects {.callout}
->
-> "NA" is a special object in R which denotes a missing value. NA can
-> occur in any type of vector. There are a few other types of
-> special objects: `Inf` denotes infinity (can be positive or negative),
-> while `NaN` means Not a number, an undefined value (i.e. `0/0`).
-> `NULL` denotes that the data structure doesn't exist (but can occur
-> in list elements).
->
-
-You can ask questions about the structure of vectors:
+You can also make series of numbers:
 
 ```{r}
-x <- 0:10
-tail(x, n=2) # get the last 'n' elements
+mySeries <- 1:10
+mySeries
+seq(10)
+seq(1,10, by=0.1)
 ```
 
-```{r}
-head(x, n=1) # get the first 'n' elements
-```
+In addition to asking for elements of a vector with the square bracket notation, we can ask a few other questions about vectors:
 
 ```{r}
+x <- seq(10)
+head(x, n=2)
+tail(x, n=4)
 length(x)
 ```
 
-```{r}
-str(x)
-```
-
-Vectors can be named:
+Finally, you can give names to elements in your vector, and ask for them that way:
 
 ```{r}
-x <- 1:4
+x <- 5:8
 names(x) <- c("a", "b", "c", "d")
 x
+x['b']
 ```
```

Comment:
As much as I love this form, I might be better to go with `x$b`. This would provide a nice callback to L40, where we talk about `cats$weight`. i.e. here is another chance to explore what the `$` operator does